In [24]:
import sqlite3
import pandas as pd

In [25]:
conn = sqlite3.connect('data/noshow.db')
df = pd.read_sql_query('SELECT * FROM noshow', conn)

In [26]:
df

no_show   branch booking_month arrival_month  arrival_day  \
0           1.0   Changi          July           May         19.0   
1           0.0  Orchard      December      February         28.0   
2           0.0  Orchard      December           May         22.0   
3           0.0  Orchard       October     September          2.0   
4           0.0  Orchard      February      February          7.0   
...         ...      ...           ...           ...          ...   
119386      0.0   Changi       October         March         17.0   
119387      0.0   Changi        August          June          2.0   
119388      0.0   Changi       January          July          1.0   
119389      0.0   Changi       January           May         26.0   
119390      0.0   Changi        August          June         12.0   

       checkout_month  checkout_day    country first_time    room  \
0                 May          20.0      China        Yes    King   
1               March          -4.0      India        Yes    None   
2                 May          24.0  Australia        Yes  Single   
3           September           4.0      China        Yes    None   
4            February           8.0      China        Yes    King   
...               ...           ...        ...        ...     ...   
119386          March          19.0      China        Yes    King   
119387           June           3.0  Indonesia        Yes    King   
119388           July           3.0  Indonesia        Yes    King   
119389            May          29.0  Indonesia        Yes    King   
119390           June          13.0      India        Yes    King   

              price platform num_adults  num_children  
0              None    Email          2           1.0  
1       USD$ 723.34    Email          1           0.0  
2       SGD$ 650.94  Website          1           0.0  
3       SGD$ 978.67    Agent          1           1.0  
4              None  Website          1           2.0  
...             ...      ...        ...           ...  
119386  SGD$ 953.94    Email          2           1.0  
119387  USD$ 636.84    Email          1           1.0  
119388         None  Website          1           0.0  
119389   USD$ 720.1    Email          1           1.0  
119390  SGD$ 898.62    Phone          1           0.0  

[119391 rows x 14 columns]

We can already see that there are is some incorrectly entered data which will need to be dealt with as we do the EDA. For example, for the second data point, the checkout day is negative. Will need to investigate if that means something. Otherwise, will have to clean it. 

We will start by checking if there are any null values

In [27]:
df.isna().sum()

no_show               1
branch                1
booking_month         1
arrival_month         1
arrival_day           1
checkout_month        1
checkout_day          1
country               1
first_time            1
room              21613
price             24882
platform              1
num_adults            1
num_children          1
dtype: int64

There seems to be one missing from almost all the columns (apart from room and price which have way more). I suspect this might be the same data point. Checking. 

In [28]:
df[df['no_show'].isna()]

no_show branch booking_month arrival_month  arrival_day  \
115536      NaN   None          None          None          NaN   

       checkout_month  checkout_day country first_time  room price platform  \
115536           None           NaN    None       None  None  None     None   

       num_adults  num_children  
115536       None           NaN

As suspected, this is just a datapoint with full null values. For ease of analysis, will just remove it now. 

In [29]:
df = df.drop(115536, axis = 0)

In [30]:
df.isna().sum()

no_show               0
branch                0
booking_month         0
arrival_month         0
arrival_day           0
checkout_month        0
checkout_day          0
country               0
first_time            0
room              21612
price             24881
platform              0
num_adults            0
num_children          0
dtype: int64

Yep so the na values for all of the features apart from room and price have been removed. Now lets take a closer look at room and price. 

In [31]:
df[df['room'].isna()]

no_show   branch booking_month arrival_month  arrival_day  \
1           0.0  Orchard      December      February         28.0   
3           0.0  Orchard       October     September          2.0   
5           0.0   Changi       January         April         17.0   
7           1.0   Changi      December         March         31.0   
8           1.0   Changi          July          June         14.0   
...         ...      ...           ...           ...          ...   
119363      0.0   Changi       October      November         19.0   
119368      1.0  Orchard         April          June          9.0   
119377      1.0  Orchard       January     September         19.0   
119379      0.0  Orchard         April          June          6.0   
119384      0.0  Orchard       October         March          9.0   

       checkout_month  checkout_day    country first_time  room         price  \
1               March          -4.0      India        Yes  None   USD$ 723.34   
3           September           4.0      China        Yes  None   SGD$ 978.67   
5               April          19.0  Indonesia        Yes  None   USD$ 659.68   
7               April           2.0      China        Yes  None   USD$ 665.39   
8                June          17.0      China        Yes  None   USD$ 700.23   
...               ...           ...        ...        ...   ...           ...   
119363       November          20.0  Indonesia        Yes  None   USD$ 587.81   
119368           June          11.0      China         No  None  SGD$ 1469.56   
119377      September          21.0      China        Yes  None   USD$ 962.33   
119379           June           7.0      Japan        Yes  None   USD$ 651.44   
119384          March          10.0  Indonesia        Yes  None  SGD$ 1245.56   

       platform num_adults  num_children  
1         Email          1           0.0  
3         Agent          1           1.0  
5         Phone          1           0.0  
7       Website          2           0.0  
8       Website        one           2.0  
...         ...        ...           ...  
119363    Email          2           1.0  
119368    Phone        two           0.0  
119377  Website          2           2.0  
119379  Website          1           0.0  
119384  Website          1           1.0  

[21612 rows x 14 columns]

It's missing a lot of data points. We will probably have to end up filling up the data as 21612 data points are missing values. Lets take a look at the missing values for price.

In [32]:
df[df['price'].isna()]

no_show   branch booking_month arrival_month  arrival_day  \
0           1.0   Changi          July           May         19.0   
4           0.0  Orchard      February      February          7.0   
19          0.0  Orchard      December       January         19.0   
27          1.0   Changi      December         March         12.0   
28          0.0   Changi          July          July         18.0   
...         ...      ...           ...           ...          ...   
119364      0.0   Changi       October       October         28.0   
119365      1.0   Changi      December           May          2.0   
119380      1.0  Orchard         March      February          2.0   
119385      1.0  Orchard          July         April         25.0   
119388      0.0   Changi       January          July          1.0   

       checkout_month  checkout_day    country first_time   room price  \
0                 May          20.0      China        Yes   King  None   
4            February           8.0      China        Yes   King  None   
19            January          20.0      China        Yes   King  None   
27              March          13.0      China        Yes   King  None   
28               July          20.0      China        Yes   King  None   
...               ...           ...        ...        ...    ...   ...   
119364        October          31.0      China        Yes   King  None   
119365            May           3.0      China        Yes   King  None   
119380       February           3.0      China        Yes   King  None   
119385          April          27.0      China        Yes  Queen  None   
119388           July           3.0  Indonesia        Yes   King  None   

       platform num_adults  num_children  
0         Email          2           1.0  
4       Website          1           2.0  
19        Agent          1           0.0  
27        Phone          1           1.0  
28        Agent          2           1.0  
...         ...        ...           ...  
119364  Website          2           1.0  
119365    Agent          2           1.0  
119380  Website          1           1.0  
119385  Website          1           0.0  
119388  Website          1           0.0  

[24881 rows x 14 columns]

I'm currently hoping that we aren't missing any values from both room and price in the same row. Lets take a look. As that would make it easier to fill in missing values. As room type would be a good predictor of price. 

In [33]:
df[df['price'].isna()&df['room'].isna()]

Empty DataFrame
Columns: [no_show, branch, booking_month, arrival_month, arrival_day, checkout_month, checkout_day, country, first_time, room, price, platform, num_adults, num_children]
Index: []